vertical-pod-autoscaler/pkg/recommender/model/cluster.go

```diff
@@ -124,18 +123,44 @@ func (cluster *ClusterState) AddOrUpdatePod(podID PodID, newLabels labels.Set, p
 		pod = newPod(podID)
 		cluster.Pods[podID] = pod
 	}
+
 	newlabelSetKey := cluster.getLabelSetKey(newLabels)
+	if podExists && pod.labelSetKey != newlabelSetKey {
+		// This Pod is already counted in the old VPA, remove the link.
+		cluster.removePodFromItsVpa(pod)
+	}
 	if !podExists || pod.labelSetKey != newlabelSetKey {
 		pod.labelSetKey = newlabelSetKey
 		// Set the links between the containers and aggregations based on the current pod labels.
 		for containerName, container := range pod.Containers {
 			containerID := ContainerID{PodID: podID, ContainerName: containerName}
 			container.aggregator = cluster.findOrCreateAggregateContainerState(containerID)
 		}
+
+		cluster.addPodToItsVpa(pod)
 	}
 	pod.Phase = phase
 }
 
+// addPodToItsVpa increases the count of Pods associated with a VPA object.
+// Does a scan similar to findOrCreateAggregateContainerState so could be optimized if needed.
+func (cluster *ClusterState) addPodToItsVpa(pod *PodState) {
+	for _, vpa := range cluster.Vpas {
+		if vpa_utils.PodLabelsMatchVPA(pod.ID.Namespace, cluster.labelSetMap[pod.labelSetKey], vpa.ID.Namespace, vpa.PodSelector) {
+			cluster.VpaPodCount[vpa.ID]++
```

Comment:
Since we are currently in the state where only one VPA object should match pod, should we return early?

Reply:
I was wondering the same and left it as it since we have the same construct in multiple places and I didn't want to end up with an outlier.
Or do you think it's still worth changing it?

Reply:
I guess it's fine then, though that other places come from the time when we had selector-based API.